Alpha/user callbacks for tracing

dialalpha opened the pull request:
<!--
*Thank you for contributing to Apache Flink Pulsar Connectors - we are happy that you want to help us improve our Flink connectors. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

## Contribution Checklist

- The name of the pull request should correspond to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
- Commits should be in the form of "[FLINK-XXXX][Component] Title of the pull request", where [FLINK-XXXX] should be replaced by the actual issue number.
    Generally, [Component] tag should indicate the part you modified. The options are: [Stream | Table | Test | E2E | Build].
    For example: "[FLINK-XXXX][Stream] XXXX" if you are working on the `DataStream` part of pulsar connector or "[FLINK-XXXX][Test] XXXX" if this pull request is only used for adding tests.
- Each pull request should only have one JIRA issue.
- Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.
-->

## Purpose of the change

*For example: Add dynamic sink topic support for Pulsar connector.*

## Brief change log

- *Change the internal design of `ProducerRegister`.*
- *Expose topic metadata query in `PulsarSinkContext`.*
- *Change the internal metadata cache in `MetadataListener`.*

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality
guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*

- *Added unit tests*
- *Added integration tests for end-to-end deployment*
- *Manually verified by running the Pulsar connector on a local Flink cluster.*

## Significant changes

*(Please check any boxes [x] if the answer is "yes". You can first publish the PR and check them afterwards, for
convenience.)*

- [ ] Dependencies have been added or upgraded
- [ ] Public API has been changed (Public API is any class annotated with `@Public(Evolving)`)
- [ ] Serializers have been changed
- [ ] New feature has been introduced
    - If yes, how is this documented? (not applicable / docs / JavaDocs / not documented)

leonxu8 commented on flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/callback/SinkUserCallback.java:
```diff
@@ -0,0 +1,36 @@
+     * @param topic the pulsar topic or partition that the message will be routed to.
+     * @return the message to send to the pulsar topic, potentially a new message.
+     */
+    default PulsarMessage<?> beforeSend(IN element, PulsarMessage<?> message, String topic) {
```
What's the scenario when users need to modify the message? If users modify the message then `element` will be inconsistent with `message`, do we need to call this out in the comment as caveat? The logic after beforeSend will still be using element.

dialalpha replied:
So this API takes inspiration from [`PulsarProducerInterceptor`](https://github.com/apache/pulsar/blob/master/pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ProducerInterceptor.java).

A scenario where we would want to modify the `message` is: let's say you want to propagate some spans, or add extra tracing headers, you could add those in the message properties. For example if you want to do tracing downstream without the cost of serialization, the next step could simply get the info it needs from the message property bag that was added in the previous step. 

Good question about the potential conflict between `element` and `message`. The spirit of this was that we don't really modify the message, but "enrich" it for downstream consumers. Open to suggestions on how to address this.

I pass the `element` in the callback so the user can avoid the cost of extra serialization if what they need for enrichment is in the message body.

dialalpha replied:
Actually, giving it a bit more thought, it would be better to just return `void` in all the callback messages and not allow mutations. The ability to modify the `message` in the send callbacks is redundant since the user already controls the serializer, and they do can whatever they want there already.

Similarly on the source side, we can make the callbacks read-only. The user can make whatever transformations they want to the message in the deserializer, or a dedicated operator. There is no need for this to be done in a callback.

syhily commented:
Hi @dialalpha @leonxu8, I have added some comments about this feature in [JIRA](https://issues.apache.org/jira/browse/FLINK-31428). You can check it. I agree with @leonxu8 that you can't modify the raw value and deserialized value in `XXCallback`. But it's ok to modify it in Pulsar's internal intercepter and PulsarXXXSchema.

dialalpha commented:
> Hi @dialalpha @leonxu8, I have added some comments about this feature in [JIRA](https://issues.apache.org/jira/browse/FLINK-31428). You can check it. I agree with @leonxu8 that you can't modify the raw value and deserialized value in XXCallback. But it's ok to modify it in Pulsar's internal intercepter and PulsarXXXSchema.

Thanks @syhily  - I added a reply to the JIRA as well. Agreed on not modifying the values - the real problem we want to avoid is the extra ser/deser if possible.
